Multi window close

Redth opened the pull request:
Adds:

- `IApplication.OnWindowClosed`

This now tracks when a UISceneSession disconnects, an Activity is destroyed, and a Window is closed and remove its Window from the Application's collection of windows.

- `IApplication.CloseWindow`

This method provides a way to request that an `IWindow`'s platform scene/activity/window is closed intentionally without the user requesting it specifically through the platform's mechanism to do so.

- [x] TODO: Windows support for detecting the window is closed
- [x] Question: Is the `OnWindowClosed` mapper the right way to do this?

Redth committed:
Use CreateWindow instead of MainPage for multiwin

If we don't use this for multi window, there may be no window after the first one / instance is closed.

Redth committed:
Update MauiUIApplicationDelegate.cs

Fix configuration name to match the key we expect.

Redth committed:
Add CloseWindow/OnWindowClosed

When a window is closed by the user (scene disconnects or activity is destroyed, or window is closed), we need to stop tracking the IWindow instance.

Also, this adds a programmatic way to request a window be closed.

mattleibow commented:
Why is the window event on the app? Maybe it should be on the window? And then maybe exposed as an event as well?

AmrAlSayed0 commented:
I agree with @mattleibow, exposing window close as an event would allow multiple subscribers instead of having to subclass. The OnWindowClose approach feels like a workaround for languages that don't have events (eg. Java) but C# does.

Also, Android's On destroy is unreliable if you want to detect if the Activity is being closed. Android calls it if there is configuration changes and it might not even be called at all if the app/activity is killed due to memory constraints.

You can call `IsChangingConfigurations` or `IsFinishing` to see if the Activity is _actually_ being closed.

Redth committed:
Remove mapper for onwindowclosed

Redth committed:
Fix window destroyed call for android

Based on feedback OnDestroy is unreliable and so using OnStop instead and checking if the activity is finished to determine if calling the xplat destroy lifecycle event is appropriate.

This should be ok even for configuration changed since if the activity isn't handling the config change, it is actually being destroyed and will undergo a full new cycle with OnCreate called again.

Redth committed:
Add some scene lifecycle events for iOS/Mac

Redth committed:
Fix up calls for closing window on iOS

Redth committed:
Remove leftover call

Redth commented:
@AmrAlSayed0 thanks for the tip on that one, changed up where we call it :)

Redth commented on src/Controls/src/Core/HandlerImpl/Application.Impl.cs:
```diff
@@ -57,6 +57,31 @@ void IApplication.OpenWindow(IWindow window)
+			if (window is null)
+				return;
+
+			if (window is NavigableElement ne)
```
@PureWeen does this seem ok?

PureWeen replied:
Looks good 👍

Redth commented on src/Core/src/Handlers/Application/ApplicationHandler.Standard.cs:
```diff
@@ -8,5 +8,7 @@ public partial class ApplicationHandler : ElementHandler<IApplication, object>
 		public static void MapTerminate(ApplicationHandler handler, IApplication application, object? args) { }
 		public static void MapOpenWindow(ApplicationHandler handler, IApplication application, object? args) { }
+		public static void MapCloseWindow(ApplicationHandler handler, IApplication application, object? args) { }
+		public static void MapOnWindowClosed(ApplicationHandler handler, IApplication application, object? args) { }
```
```suggestion
```

Redth commented on src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.iOS.cs:
```diff
@@ -36,6 +37,13 @@ static void OnConfigureLifeCycle(IiOSLifecycleBuilder iOS)
 				.WillTerminate(app =>
 				{
 					app.GetWindow()?.Destroying();
```
```suggestion
					// By this point if we were a multi window app, the GetWindow would be null anyway
					app.GetWindow()?.Destroying();
```

Redth commented on src/Core/src/Platform/Android/MauiAppCompatActivity.cs:
```diff
@@ -1,5 +1,6 @@
 using Android.OS;
 using AndroidX.AppCompat.App;
+using Microsoft.Maui.LifecycleEvents;
```
```suggestion
```

Redth committed:
Update src/Core/src/Handlers/Application/ApplicationHandler.Standard.cs

Redth committed:
Update src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.iOS.cs

Redth committed:
Update src/Core/src/Platform/Android/MauiAppCompatActivity.cs

mattleibow commented on src/Core/src/Handlers/Application/ApplicationHandler.iOS.cs:
```diff
@@ -23,6 +23,24 @@ public static void MapOpenWindow(ApplicationHandler handler, IApplication applic
+			if (args is IWindow window)
+			{
+				// See if the window's handler has an associated UIWindowScene and UISceneSession
+				var sceneSession = window.Handler?.MauiContext?.GetNativeWindow()?.WindowScene?.Session;
```
window.Handler.NativeView should be UIWindow... Can you use that?

Redth committed:
Implement close on Windows

Redth committed:
Simplify how we get the scene from a window

Redth committed:
Add a public method for CloseWindow in Application for Controls impl

Redth committed:
Add a method to get the parent from a page

Redth committed:
Bad logic fix

Redth committed:
Fix logic for using scene vs not

Redth committed:
Merge branch 'main' into multi-window-close

mattleibow reviewed (approved):
Feels good. Besides the nav area that @PureWeen should give an eyeball to just to be safe.

Redth committed:
Add close button to window sample

mattleibow committed:
fix test stubs

mattleibow committed:
Merge remote-tracking branch 'origin/main' into multi-window-close

mattleibow committed:
Fix more tests
